File: sol/1074.py
class Solution:
    def numSubmatrixSumTarget(self, matrix: List[List[int]], target: int) -> int:
        n = len(matrix)
        m = len(matrix[0])
        for j in range(1, m):
            matrix[0][j] += matrix[0][j - 1]
        for i in range(1, n):
            matrix[i][0] += matrix[i - 1][0]
            for j in range(1, m):
                matrix[i][j] += matrix[i][j - 1] + matrix[i - 1][j] - matrix[i - 1][j - 1]
        count = 0
        for r1 in range(n):
            for r2 in range(r1, n):
                rec = {0: 1}
                for c in range(m):
                    sub = matrix[r2][c] - (matrix[r1 - 1][c] if r1 > 0 else 0)
                    count += rec.get(sub - target, 0)
                    rec[sub] = rec.get(sub, 0) + 1
        return count

File: sol/1074/1074.cpp
#include <unordered_map>

struct Solution {
    int numSubmatrixSumTarget(vector<vector<int>>& matrix, int target) {
       int n = matrix.size(), m = matrix[0].size();
       for (int j = 1; j < m; ++j) {
           matrix[0][j] += matrix[0][j - 1];
       }
       for (int i = 1; i < n; ++i){
           matrix[i][0] += matrix[i - 1][0];
           for (int j = 1; j < m; ++j){
               matrix[i][j] += matrix[i][j - 1] - matrix[i - 1][j - 1] + matrix[i - 1][j];
           }
       }
       int count = 0, sub;
       unordered_map<int, int> rec;
       for (int r1 = 0; r1 < n; ++r1){
           for (int r2 = r1; r2 < n; ++r2){
               rec.clear();
               rec.insert({0, 1});
               for (int c = 0; c < m; ++c){
                   sub = matrix[r2][c] - (r1 > 0 ? matrix[r1 - 1][c] : 0);
                   count += rec[sub - target];
                   ++rec[sub];
               }
           }
       }
       return count;
    }
};
